sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -1282,6 +1283,14 @@ private static Parameter analyzeExtraParameter(
           rawType.equals(Instant.class),
           "@Timestamp argument must have type org.joda.time.Instant.");
       return Parameter.timestampParameter();
+    } else if (hasAnnotation(DoFn.KeyId.class, param.getAnnotations())) {
+      Type keyType = ((ParameterizedType) inputT.getType()).getActualTypeArguments()[0];
```

Comment:
I think we also need to verify that the input type is a KV

Reply:
Added with a test case